ops/model.py

```diff
@@ -3596,3 +3621,60 @@ def _ensure_loaded(self):
         self._notice = self._container.get_notice(self.id)
         assert self._notice.type == self.type
         assert self._notice.key == self.key
+
+
+class CloudSpec:
+    """Cloud specification information (metadata) including credentials."""
+
+    def __init__(self,
+                 type: str,
+                 name: str,
+                 region: Optional[str],
+                 endpoint: Optional[str],
+                 is_controller_cloud: Optional[str],
+                 credential: Optional[Dict[str, Any]],
```

Comment:
I had suggested to make this a new `CloudCredential` dataclass, as it has a few other fields than just the credential attributes. See my comments about "nested object" here: https://github.com/canonical/operator/issues/186#issuecomment-1998442238 ... any reason we shouldn't do that to match the "shape" of the Juju object?

Reply:
I did not fully copy from Juju because I'm a bit confused here: In Juju [`CloudCredential.Attributes` reads from json `attrs`](https://github.com/juju/juju/blob/3.5/rpc/params/cloud.go#L153), but in [AWS charm](https://github.com/Automatic/charm-aws-integrator/blob/a9ab3729107debb220293187970c74f5af80e69b/lib/charms/layer/aws.py#L59) and [Azure charm](https://github.com/phvalguima/charm-azure-integrator/blob/479bacacb0762128cd1b667acf78d3eb5e8b4883/lib/charms/layer/azure.py#L66) it seems they are reading from "attributes" instead of "attrs". I do not know if the cloud credentials in other clouds are of different structure, so to be on the safer end I used a dictionary for it.

Should I create an object CloudCredential to match Juju anyway?

Reply:
The Juju .go code is almost certainly the source of truth, because that's what's in Juju. Maybe in older versions the JSON field was called "attributes"? I'm not sure why the discrepancy between Juju and what the charm is doing. Can you see what shape of object / field names the current Juju version returns?

Reply:
I did some testing on AWS EC2 + LXD:

If I run juju CLI directly to show-crendential the result structure is quite diffrent, like:

```bash
$ juju show-credential aws default
client-credentials:
  aws:
    default:
      content:
        auth-type: access-key
        access-key: xxx
```

If I run the hook tool, what I got is similar as what the AWS and Azure charms have:

```json
credential:
  authtype: certificate
  attributes:
    client-cert: foo
    client-key: bar
    server-cert: baz
  redacted: []
```

Does this mean we should create a `CloudCredential` dataclass, but instead of using `attrs`, we use `attributes`?

Reply:
Aha, I've figured out the discrepancy: the default formatting is YAML, and those structs don't have yaml tags, so the yaml marshalling uses the default field name (lowercased version of the Go field name in the struct). So let's use `credential-get --format=json` -- you can call `_ModelBackend._run(use_json=True)` to do that -- and that way we'll get the explicitly defined JSON field names: auth-type, attrs, redacted, etc.

So yes, use `--format=json`, copy from the Juju struct fields json:"..." field name, and add CloudSpec and CloudCredential dataclasses.

Reply:
Got it. It's using `--format=json` already, see[here](https://github.com/canonical/operator/pull/1152/files#diff-b7713b8731a4d1f895063e58f31f8ab54f7bb6ca0e2b1712d85683874d94a6e6R3862).

I'll mark it as resolved for now.